### PR TITLE
New debugscreen implementation

### PR DESCRIPTION
**Pull request by windsurfer1122**
See also https://github.com/vitasdk/samples/issues/50

debugScreen
-----------
* Splitted code and shared definitions into .c and .h file, so it can be used in bigger projects with multiple source files.
  Added custom header file to avoid people messing up the implementation in their projects.
* Fixed CSI 0J command. Enhanced sample debugscreen cases to show the issue.
* Added documentation about CSI commands at top and comments in C functions.
* Implemented ANSI/VTERM/GREYSCALE colors via lookup table for correct colors.
* Implemented color state machine to comply better with ECMA-48 / ISO/IEC 6429:1992 and Graphic Rendition Combination Mode (GRCM) "Cumulative".
* Added CSI 22m. Set standard intensity ("normal" color). Accomplishes CSI 1m and 2m.
* Added CSI 7m and 27m. Enable/disable FG/BG inversion.
* Support 16 storage slots for CSI s and u.
* Enhanced bitmap routines to work also with not byte-aligned glyphs.
  * Enhanced HTML bitmap editor to work also with not byte-aligned glyphs.
* Draw dummy glyph (dotted line in the middle) in correct font size for not available glyphs.
* Added function to get a copy of current color state.
* Added functions to get/set pixel coordinates for custom save storages and custom positioning.
* Added functions to get/set font.
* Added function to scale a font by 2.
* Sanitized macro usage.
* Fixed typos.
* Extended sample "debugscreen" to test more cases and demonstrate new functions:
  * used different background colors to better show CSI J and K commands
  * added font scaling
  * added inversion on/off via CSI 7m and CSI 27m
  * show "dark" white on "bright" black
  * added missing glyph
  * show all colors CSI 30-37/40-47/90-97/100-107 with their RGB code
* Tested nearly all samples, except for network samples.

Sample camera
-------------
* Added missing include <psp2/kernel/processmgr.h> for sceKernelExitProcess().

Sample microphone
-----------------
* Added missing include <string.h> for memset() and memcpy().
* Changed direct access to coordX/Y to CSI sequences.

Sample net_http_bsd
-------------------
* Changed direct access to font to new functions.

Sample power
------------
* Added missing include <stdio.h> for FILE, fopen(), fclose().
* Added missing include <string.h> for memset().

### CODE REVIEW

**Comment by windsurfer1122** [comment]
0J issue with old implementation
---
debugscreen_old_sample_0J_issue

New debugscreen sample
---
debugscreen_sample_20190725

Tests with sample `net_http_bsd`
---
net_http_bsd_wttr in
net_http_bsd_rate sx

**Comment by windsurfer1122** [comment]
Corrected header definition for C++.
No need for previous change to cpp files.

File: common/debugScreen_custom.h [hunk 1]
```diff
@@ -0,0 +1,17 @@
+#ifndef DEBUG_SCREEN_CUSTOM_H
+#define DEBUG_SCREEN_CUSTOM_H
+
+//#define SCREEN_TAB_SIZE (8)
+
+// backward compatibility for sources based on older Vita SDK versions
+#define DEBUG_SCREEN_CODE_INCLUDE // not recommended for your own projects, but for sake of backward compatibility
```
Comment (yne) [inline common/debugScreen_custom.h]: > not recommended for your own projects, but for sake of backward compatibility

yet you use them in the demo sample.
Why not `#define` it just in the sample that need them ?

**Comment by windsurfer1122** [comment]
I was thinking about people that already use the previous debugscreen implementation, so they can just use the new one as a simple drop-in.

No sample of Vita SDK needs that inclusion. But the CMakeLists.txt of all samples would have to be changed adding "../common/debugscreen.c" to the executable.

**Comment by windsurfer1122** [comment]
Checked failed due to an curl error. How to re-trigger the test build?

**Comment by d3m3vilurr** [comment]
does it ready to merge?

**Comment by yne** [comment]
Sounds good here

**Comment by windsurfer1122** [comment]
From my point of view it is ready for merge - but I'm not a VitaSDK member.

**Comment by d3m3vilurr** [comment]
Thank you for your hard work. :)

**Pull request merged**